src/aspects/ec2-automated-shutdown/index.ts

```diff
@@ -0,0 +1,142 @@
+import { LambdaConfiguration } from '../../interfaces/lambda-configuration';
+
+export interface EC2ShutdownProps {
+    /**
+     * The % CPU threshold below which the instance will be shut down
+     */
+    readonly cpuThreshold: number;
+
+    /**
+     * Optional Lambda configuration settings.
+     */
+    readonly lambdaConfiguration?: LambdaConfiguration;
+
+    /**
+     * Optional KMS Encryption Key to use for encrypting resources.
+     */
+    readonly encryption?: IKey;
+}
+
+/**
+ * Aspect that applies a mechanism to automatically shut down
+ * an EC2 instance when its CPU utilization falls below a specified threshold.
+ * Allows for cost optimization and the reduction of resources not being actively used.
+ */
+export class Ec2AutomatedShutdown implements IAspect {
+    private lambdaFunction!: SecureFunction;
+    private processedStacks: Set<Stack> = new Set();
+    private counter = 0;
+
+    constructor(private readonly props: EC2ShutdownProps) {
+        if (props.cpuThreshold <= 0 || props.cpuThreshold > 100) {
+            throw new Error('CPU threshold must be between 0 and 100');
+        }
+    }
+
+    visit(node: IConstruct): void {
+        if (!(node instanceof Instance) && !(node instanceof CfnInstance)) {
+            return;
+        }
+
+        const stack = Stack.of(node);
+
+        if (!this.processedStacks.has(stack)) {
+            this.ensureLambdaFunction(stack);
+            this.processedStacks.add(stack);
+        }
+
+        this.createShutdownMechanism(node);
+    }
+
+    private ensureLambdaFunction(stack: Stack): void {
+        if (!this.lambdaFunction) {
+            this.lambdaFunction = new SecureFunction(
+                stack,
+                'EC2ShutdownFunction',
+                {
+                    code: Code.fromAsset(join(__dirname, 'handler')),
+                    runtime: Runtime.NODEJS_20_X,
+                    handler: 'index.handler',
+                    timeout: Duration.seconds(15),
+                    encryption: this.props.encryption,
+                    ...this.props.lambdaConfiguration
+                }
+            );
+
+            this.lambdaFunction.role.addToPolicy(
+                new PolicyStatement({
+                    actions: ['ec2:StopInstances'],
+                    resources: ['*']
+                })
+            );
+        }
+    }
+
+    private createShutdownMechanism(instance: IConstruct): void {
+        if (!this.lambdaFunction) {
+            throw new Error('Lambda Function not initialized');
+        }
+
+        let instanceId: string;
+        if (instance instanceof Instance) {
+            instanceId = instance.instanceId;
+        } else if (instance instanceof CfnInstance) {
+            instanceId = instance.ref;
+        } else {
+            throw new Error('Unsupported instance type');
+        }
+
+        const uniqueId = `${++this.counter}`;
+        const topic = new Topic(instance, `ShutdownTopic-${uniqueId}}`);
```

Comment:
I would prefer we include the name of the Ec2 that this is applying to in the name of supporting resources. That way it is easier for the user to see what resources are being used to support the shutdown for each instance.